src/client/websocket/WebSocketConnection.js

```diff
@@ -26,91 +28,391 @@ const WebSocket = (function findWebSocket() {
  */
 class WebSocketConnection extends EventEmitter {
   /**
+   * @param {WebSocketManager} manager the WebSocket manager
    * @param {string} gateway Websocket gateway to connect to
    */
-  constructor(gateway) {
-    super(gateway);
-    this.ws = new WebSocket(gateway);
-    if (browser) this.ws.binaryType = 'arraybuffer';
-    this.ws.onmessage = this.eventMessage.bind(this);
-    this.ws.onopen = this.emit.bind(this, 'open');
-    this.ws.onclose = this.emit.bind(this, 'close');
-    this.ws.onerror = this.emit.bind(this, 'error');
+  constructor(manager, gateway) {
+    super();
+    /**
+     * WebSocket Manager of this connection
+     * @type {WebSocketManager}
+     */
+    this.manager = manager;
+    /**
+     * Client this belongs to
+     * @type {Client}
+     */
+    this.client = manager.client;
+    /**
+     * WebSocket connection itself
+     * @type {WebSocket}
+     */
+    this.ws = null;
+    /**
+     * Current sequence of the WebSocket
+     * @type {number}
+     */
+    this.sequence = -1;
+    /**
+     * Current status of the client
+     * @type {number}
+     */
+    this.status = Constants.Status.IDLE;
+    /**
+     * Packet Manager of the connection
+     * @type {WebSocketPacketManager}
+     */
+    this.packetManager = new PacketManager(this);
+    /**
+     * Last time a ping was sent (a timestamp)
+     * @type {number}
+     */
+    this.pingSendTime = 0;
+    /**
+     * Contains the rate limit queue and metadata
+     * @type {Object}
+     */
+    this.rateLimit = {
+      queue: [],
+      remaining: 120,
+      resetTime: -1,
+    };
+    this.connect(gateway);
   }
 
   /**
-   * Called when the websocket gets a message
-   * @param {Object} event Close event object
-   * @returns {Promise<boolean>}
+   * Causes the client to be marked as ready and emits the ready event
+   * @returns {void}
    */
-  eventMessage(event) {
-    try {
-      const data = this.unpack(event.data);
-      this.emit('packet', data);
-      return true;
-    } catch (err) {
-      if (this.listenerCount('decodeError')) this.emit('decodeError', err);
-      return false;
-    }
+  triggerReady() {
+    if (this.status === Constants.Status.READY) return this.debug('Tried to mark self as ready, but already ready');
+    this.status = Constants.Status.READY;
+    this.client.emit(Constants.Events.READY);
+    return this.packetManager.handleQueue();
   }
 
   /**
-   * Send data over the websocket
-   * @param {string|Buffer} data Data to send
+   * Checks whether the client is ready to be marked as ready
+   * @returns {void}
    */
-  send(data) {
-    this.ws.send(this.pack(data));
+  checkIfReady() {
+    if (this.status === Constants.Status.READY || this.status === Constants.Status.NEARLY) return false;
+    let unavailableGuilds = 0;
+    for (const guild of this.client.guilds.values()) {
+      unavailableGuilds += guild.available ? 0 : 1;
```

Comment:
`if (!guild.available) unavailableGuilds++` will be faster if the guild is available